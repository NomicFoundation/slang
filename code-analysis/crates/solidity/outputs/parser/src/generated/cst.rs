// This file is generated via `cargo build`. Please don't edit by hand.

use super::kinds;
use super::parse::Context;
use serde::Serialize;
use std::ops::Range;
#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub enum Node<'a> {
    None,
    Rule {
        kind: kinds::Rule,
        children: Vec<&'a Node<'a>>,
    },
    Token {
        kind: kinds::Token,
        #[doc = r" Range doesn't include the trivia"]
        range: Range<usize>,
        #[doc = r" Only Trivia"]
        #[serde(skip_serializing_if = "Vec::is_empty")]
        trivia: Vec<&'a Node<'a>>,
    },
    #[doc = r" For anonymous groups referenced from AST nodes i.e. `delimited_by`"]
    Group {
        #[serde(skip_serializing_if = "Vec::is_empty")]
        children: Vec<&'a Node<'a>>,
    },
}
#[allow(unused_variables)]
pub trait Visitor {
    fn enter_rule(
        &mut self,
        kind: kinds::Rule,
        children: &Vec<&Node<'_>>,
        node: &Node<'_>,
    ) -> VisitorEntryResponse {
        VisitorEntryResponse::StepIn
    }
    fn exit_rule(
        &mut self,
        kind: kinds::Rule,
        children: &Vec<&Node<'_>>,
        node: &Node<'_>,
    ) -> VisitorExitResponse {
        VisitorExitResponse::StepIn
    }
    fn enter_token(
        &mut self,
        kind: kinds::Token,
        range: &Range<usize>,
        trivia: &Vec<&Node<'_>>,
        node: &Node<'_>,
    ) -> VisitorEntryResponse {
        VisitorEntryResponse::StepIn
    }
    fn exit_token(
        &mut self,
        kind: kinds::Token,
        range: &Range<usize>,
        trivia: &Vec<&Node<'_>>,
        node: &Node<'_>,
    ) -> VisitorExitResponse {
        VisitorExitResponse::StepIn
    }
    fn enter_group(&mut self, children: &Vec<&Node<'_>>, node: &Node<'_>) -> VisitorEntryResponse {
        VisitorEntryResponse::StepIn
    }
    fn exit_group(&mut self, children: &Vec<&Node<'_>>, node: &Node<'_>) -> VisitorExitResponse {
        VisitorExitResponse::StepIn
    }
    fn visit_none(&mut self, node: &Node<'_>) -> VisitorExitResponse {
        VisitorExitResponse::StepIn
    }
}
pub enum VisitorEntryResponse {
    Quit,
    StepIn,
    StepOver,
}
pub enum VisitorExitResponse {
    Quit,
    StepIn,
}
pub trait Visitable {
    fn visit<T: Visitor>(&self, visitor: &mut T) -> VisitorExitResponse;
}
impl<'a> Node<'a> {
    #[inline]
    pub fn none(context: &'a Context<'a>) -> &'a Node<'a> {
        context.alloc_cst_node(Self::None)
    }
    #[inline]
    pub fn rule(
        context: &'a Context<'a>,
        kind: kinds::Rule,
        children: Vec<&'a Node<'a>>,
    ) -> &'a Node<'a> {
        context.alloc_cst_node(Self::Rule { kind, children })
    }
    #[inline]
    pub fn trivia_token(
        context: &'a Context<'a>,
        range: Range<usize>,
        kind: kinds::Token,
    ) -> &'a Node<'a> {
        context.alloc_cst_node(Self::Token {
            range,
            kind,
            trivia: vec![],
        })
    }
    #[inline]
    pub fn token(
        context: &'a Context<'a>,
        range: Range<usize>,
        kind: kinds::Token,
        leading_trivia: &'a Node<'a>,
        trailing_trivia: &'a Node<'a>,
    ) -> &'a Node<'a> {
        let mut trivia = vec![];
        if *leading_trivia != Node::None {
            trivia.push(leading_trivia)
        }
        if *trailing_trivia != Node::None {
            trivia.push(trailing_trivia)
        }
        context.alloc_cst_node(Self::Token {
            range,
            kind,
            trivia,
        })
    }
    #[inline]
    pub fn group(context: &'a Context<'a>, children: Vec<&'a Node<'a>>) -> &'a Node<'a> {
        if children.is_empty() {
            Self::none(context)
        } else {
            context.alloc_cst_node(Self::Group { children })
        }
    }
}
impl<'a> Visitable for &'a Node<'a> {
    fn visit<T: Visitor>(&self, visitor: &mut T) -> VisitorExitResponse {
        match self {
            Node::None => visitor.visit_none(self),
            Node::Rule { kind, children } => {
                match visitor.enter_rule(*kind, children, self) {
                    VisitorEntryResponse::Quit => return VisitorExitResponse::Quit,
                    VisitorEntryResponse::StepOver => {}
                    VisitorEntryResponse::StepIn => {
                        for child in children {
                            match child.visit(visitor) {
                                VisitorExitResponse::Quit => return VisitorExitResponse::Quit,
                                VisitorExitResponse::StepIn => {}
                            }
                        }
                    }
                }
                visitor.exit_rule(*kind, children, self)
            }
            Node::Token {
                kind,
                range,
                trivia,
            } => {
                match visitor.enter_token(*kind, range, trivia, self) {
                    VisitorEntryResponse::Quit => return VisitorExitResponse::Quit,
                    VisitorEntryResponse::StepOver => {}
                    VisitorEntryResponse::StepIn => {
                        for child in trivia {
                            match child.visit(visitor) {
                                VisitorExitResponse::Quit => return VisitorExitResponse::Quit,
                                VisitorExitResponse::StepIn => {}
                            }
                        }
                    }
                }
                visitor.exit_token(*kind, range, trivia, self)
            }
            Node::Group { children } => {
                match visitor.enter_group(children, self) {
                    VisitorEntryResponse::Quit => return VisitorExitResponse::Quit,
                    VisitorEntryResponse::StepOver => {}
                    VisitorEntryResponse::StepIn => {
                        for child in children {
                            match child.visit(visitor) {
                                VisitorExitResponse::Quit => return VisitorExitResponse::Quit,
                                VisitorExitResponse::StepIn => {}
                            }
                        }
                    }
                }
                visitor.exit_group(children, self)
            }
        }
    }
}
