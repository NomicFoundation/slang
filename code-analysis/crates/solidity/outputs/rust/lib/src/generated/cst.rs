// This file is generated automatically by infrastructure scripts. Please don't edit by hand.

use super::kinds;
use super::lex;
use serde::Serialize;
use std::rc::Rc;
#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub enum Node {
    Rule {
        kind: kinds::Rule,
        children: Vec<Rc<Node>>,
    },
    Token {
        kind: kinds::Token,
        lex_node: Rc<lex::Node>,
        #[serde(skip_serializing_if = "Vec::is_empty")]
        trivia: Vec<Rc<Node>>,
    },
    #[doc = r" For anonymous groups referenced from AST nodes i.e. `delimited_by`"]
    Group { children: Vec<Rc<Node>> },
}
impl Node {
    pub fn rule(kind: kinds::Rule, children: Vec<Option<Rc<Self>>>) -> Option<Rc<Self>> {
        let children: Vec<_> = children.into_iter().filter_map(|e| e).collect();
        if children.is_empty() {
            None
        } else {
            Some(Rc::new(Self::Rule { kind, children }))
        }
    }
    pub fn trivia_token(kind: kinds::Token, lex_node: Rc<lex::Node>) -> Option<Rc<Self>> {
        Some(Rc::new(Self::Token {
            kind,
            lex_node,
            trivia: vec![],
        }))
    }
    pub fn token(
        kind: kinds::Token,
        lex_node: Rc<lex::Node>,
        leading_trivia: Option<Rc<Self>>,
        trailing_trivia: Option<Rc<Self>>,
    ) -> Option<Rc<Self>> {
        let mut trivia = vec![];
        if let Some(leading_trivia) = leading_trivia {
            trivia.push(leading_trivia)
        }
        if let Some(trailing_trivia) = trailing_trivia {
            trivia.push(trailing_trivia)
        }
        Some(Rc::new(Self::Token {
            kind,
            lex_node,
            trivia,
        }))
    }
    pub fn group(children: Vec<Option<Rc<Self>>>) -> Option<Rc<Self>> {
        let children: Vec<_> = children.into_iter().filter_map(|e| e).collect();
        if children.is_empty() {
            None
        } else {
            Some(Rc::new(Self::Group { children }))
        }
    }
    pub fn top_level_token(lex_node: Option<Rc<lex::Node>>) -> Rc<Self> {
        if let Some(lex_node) = lex_node {
            if let lex::Node::Named(kind, lex_node) = lex_node.as_ref() {
                Rc::new(Self::Token {
                    kind: *kind,
                    lex_node: lex_node.clone(),
                    trivia: vec![],
                })
            } else {
                unreachable!("Top level token unexpected result: {:?}", lex_node)
            }
        } else {
            unreachable!("Top level token unexpected None")
        }
    }
    pub fn top_level_rule(kind: kinds::Rule, node: Option<Rc<Self>>) -> Rc<Self> {
        node.unwrap_or_else(|| {
            Rc::new(Self::Rule {
                kind,
                children: vec![],
            })
        })
    }
}
#[allow(unused_variables)]
pub trait Visitor<E> {
    fn enter_rule(
        &mut self,
        kind: kinds::Rule,
        children: &Vec<Rc<Node>>,
        node: &Rc<Node>,
        path: &Vec<Rc<Node>>,
    ) -> Result<VisitorEntryResponse, E> {
        Ok(VisitorEntryResponse::StepIn)
    }
    fn exit_rule(
        &mut self,
        kind: kinds::Rule,
        children: &Vec<Rc<Node>>,
        node: &Rc<Node>,
        path: &Vec<Rc<Node>>,
    ) -> Result<VisitorExitResponse, E> {
        Ok(VisitorExitResponse::StepIn)
    }
    fn enter_token(
        &mut self,
        kind: kinds::Token,
        lex_node: &Rc<lex::Node>,
        trivia: &Vec<Rc<Node>>,
        node: &Rc<Node>,
        path: &Vec<Rc<Node>>,
    ) -> Result<VisitorEntryResponse, E> {
        Ok(VisitorEntryResponse::StepIn)
    }
    fn exit_token(
        &mut self,
        kind: kinds::Token,
        lex_node: &Rc<lex::Node>,
        trivia: &Vec<Rc<Node>>,
        node: &Rc<Node>,
        path: &Vec<Rc<Node>>,
    ) -> Result<VisitorExitResponse, E> {
        Ok(VisitorExitResponse::StepIn)
    }
    fn enter_group(
        &mut self,
        children: &Vec<Rc<Node>>,
        node: &Rc<Node>,
        path: &Vec<Rc<Node>>,
    ) -> Result<VisitorEntryResponse, E> {
        Ok(VisitorEntryResponse::StepIn)
    }
    fn exit_group(
        &mut self,
        children: &Vec<Rc<Node>>,
        node: &Rc<Node>,
        path: &Vec<Rc<Node>>,
    ) -> Result<VisitorExitResponse, E> {
        Ok(VisitorExitResponse::StepIn)
    }
}
pub enum VisitorEntryResponse {
    Quit,
    StepIn,
    StepOver,
}
pub enum VisitorExitResponse {
    Quit,
    StepIn,
}
pub trait Visitable<T: Visitor<E>, E> {
    fn visit(&self, visitor: &mut T) -> Result<VisitorExitResponse, E>;
    fn visit_with_path(
        &self,
        visitor: &mut T,
        path: &mut Vec<Rc<Node>>,
    ) -> Result<VisitorExitResponse, E>;
}
impl<T: Visitor<E>, E> Visitable<T, E> for Rc<Node> {
    fn visit(&self, visitor: &mut T) -> Result<VisitorExitResponse, E> {
        self.visit_with_path(visitor, &mut Vec::new())
    }
    fn visit_with_path(
        &self,
        visitor: &mut T,
        path: &mut Vec<Rc<Node>>,
    ) -> Result<VisitorExitResponse, E> {
        match self.as_ref() {
            Node::Rule { kind, children } => {
                match visitor.enter_rule(*kind, children, self, path)? {
                    VisitorEntryResponse::Quit => return Ok(VisitorExitResponse::Quit),
                    VisitorEntryResponse::StepOver => {}
                    VisitorEntryResponse::StepIn => {
                        path.push(self.clone());
                        for child in children {
                            match child.visit_with_path(visitor, path)? {
                                VisitorExitResponse::Quit => {
                                    path.pop();
                                    return Ok(VisitorExitResponse::Quit);
                                }
                                VisitorExitResponse::StepIn => {}
                            }
                        }
                        path.pop();
                    }
                }
                visitor.exit_rule(*kind, children, self, path)
            }
            Node::Token {
                kind,
                lex_node,
                trivia,
            } => {
                match visitor.enter_token(*kind, lex_node, trivia, self, path)? {
                    VisitorEntryResponse::Quit => return Ok(VisitorExitResponse::Quit),
                    VisitorEntryResponse::StepOver => {}
                    VisitorEntryResponse::StepIn => {
                        path.push(self.clone());
                        for child in trivia {
                            match child.visit_with_path(visitor, path)? {
                                VisitorExitResponse::Quit => {
                                    path.pop();
                                    return Ok(VisitorExitResponse::Quit);
                                }
                                VisitorExitResponse::StepIn => {}
                            }
                        }
                        path.pop();
                    }
                }
                visitor.exit_token(*kind, lex_node, trivia, self, path)
            }
            Node::Group { children } => {
                match visitor.enter_group(children, self, path)? {
                    VisitorEntryResponse::Quit => return Ok(VisitorExitResponse::Quit),
                    VisitorEntryResponse::StepOver => {}
                    VisitorEntryResponse::StepIn => {
                        path.push(self.clone());
                        for child in children {
                            match child.visit_with_path(visitor, path)? {
                                VisitorExitResponse::Quit => {
                                    path.pop();
                                    return Ok(VisitorExitResponse::Quit);
                                }
                                VisitorExitResponse::StepIn => {}
                            }
                        }
                        path.pop();
                    }
                }
                visitor.exit_group(children, self, path)
            }
        }
    }
}
