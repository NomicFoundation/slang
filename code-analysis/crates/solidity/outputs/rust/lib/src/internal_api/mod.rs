// Internal API will be removed/refactored along https://github.com/NomicFoundation/slang/issues/195

pub mod parser;
