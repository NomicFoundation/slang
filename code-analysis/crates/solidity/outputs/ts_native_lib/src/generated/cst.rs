// This file is generated via `cargo build`. Please don't edit by hand.

use super::kinds;
use serde::Serialize;
use std::ops::Range;
use std::rc::Rc;
#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub enum Node {
    None,
    Rule {
        kind: kinds::Rule,
        #[serde(skip_serializing_if = "Vec::is_empty")]
        children: Vec<Rc<Node>>,
    },
    Token {
        kind: kinds::Token,
        #[doc = r" Range doesn't include the trivia"]
        range: Range<usize>,
        #[doc = r" Only Trivia"]
        #[serde(skip_serializing_if = "Vec::is_empty")]
        trivia: Vec<Rc<Node>>,
    },
    #[doc = r" For anonymous groups referenced from AST nodes i.e. `delimited_by`"]
    Group {
        #[serde(skip_serializing_if = "Vec::is_empty")]
        children: Vec<Rc<Node>>,
    },
}
