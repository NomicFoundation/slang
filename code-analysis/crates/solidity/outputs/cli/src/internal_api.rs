use ariadne::{Color, Config, Fmt, Label, Report, ReportKind, Source};
use chumsky::{Parser, Span};

use solidity_rust_lib::generated::parse::{ErrorType, ParserType, SpanType};

pub struct ParserOutput<TNode> {
    pub root_node: Option<TNode>,
    pub error_reports: Vec<String>,
}

pub fn parse<'a, TNode>(
    source: &'a str,
    parser: ParserType<'a, TNode>,
    with_color: bool,
) -> ParserOutput<TNode> {
    let (parse_tree, errors) = parser.parse_recovery(source);

    let mut error_reports = vec![];
    for error in &errors {
        let source = Source::from(source);
        let report = render_error_report(&error, with_color);

        let mut result = vec![];
        report
            .write(source, &mut result)
            .expect("Failed to write report");

        let result = String::from_utf8(result).expect("Failed to convert report to utf8");
        error_reports.push(result);
    }

    return ParserOutput {
        root_node: parse_tree,
        error_reports,
    };
}

fn render_error_report<'a>(error: &ErrorType, with_color: bool) -> Report<SpanType> {
    let builder = Report::build(
        ReportKind::Error,
        error.span().context(),
        error.span().start(),
    );

    let builder = builder.with_config(Config::default().with_color(with_color));

    // TODO(OmarTawfik): also disable colors below if with_color is false

    let msg = if let chumsky::error::SimpleReason::Custom(msg) = error.reason() {
        msg.clone()
    } else {
        format!(
            "{}{}, expected {}",
            if error.found().is_some() {
                "Unexpected token"
            } else {
                "Unexpected end of input"
            },
            if let Some(label) = error.label() {
                format!(" while parsing {}", label)
            } else {
                String::new()
            },
            if error.expected().len() == 0 {
                "something else".to_string()
            } else {
                error
                    .expected()
                    .map(|expected| match expected {
                        Some(expected) => expected.to_string(),
                        None => "end of input".to_string(),
                    })
                    .collect::<Vec<_>>()
                    .join(" or ")
            },
        )
    };

    let builder = builder.with_code(3).with_message(msg).with_label(
        Label::new(error.span())
            .with_message(match error.reason() {
                chumsky::error::SimpleReason::Custom(msg) => msg.clone(),
                _ => format!(
                    "Unexpected {}",
                    error
                        .found()
                        .map(|c| format!("token {}", c.fg(Color::Red)))
                        .unwrap_or_else(|| "end of input".to_string())
                ),
            })
            .with_color(Color::Red),
    );

    let builder = match error.reason() {
        chumsky::error::SimpleReason::Unclosed { span, delimiter } => builder.with_label(
            Label::new(span.clone())
                .with_message(format!(
                    "Unclosed delimiter {}",
                    delimiter.fg(Color::Yellow)
                ))
                .with_color(Color::Yellow),
        ),
        chumsky::error::SimpleReason::Unexpected => builder,
        chumsky::error::SimpleReason::Custom(_) => builder,
    };

    return builder.finish();
}
