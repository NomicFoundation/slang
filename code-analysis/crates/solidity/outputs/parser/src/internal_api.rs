use ariadne::{Color, Config, Fmt, Label, Report, ReportKind, Source};
use chumsky::{Parser, Span, Stream};

use crate::generated::parse::{Context, ErrorType, ParserType, SpanType};

pub struct ParserOutput<TNode> {
    pub root_node: Option<TNode>,
    pub error_reports: Vec<String>,
}

pub fn parse<'a, TNode>(
    context: &'a Context<'a>,
    source: &'a str,
    parser: ParserType<'a, TNode>,
    with_color: bool,
) -> ParserOutput<TNode> {
    let input = Input::new(source, context);
    let stream = Stream::from_iter(input.eos(), input);

    let (parse_tree, errors) = parser.parse_recovery(stream);

    let mut error_reports = vec![];
    for error in &errors {
        let source = Source::from(source);
        let context = error.span().context();
        let report = render_error_report(&error, with_color);

        let mut result = vec![];
        report
            .write((context, source), &mut result)
            .expect("Failed to write report");

        let result = String::from_utf8(result).expect("Failed to convert report to utf8");
        error_reports.push(result);
    }

    return ParserOutput {
        root_node: parse_tree,
        error_reports,
    };
}

struct Input<'a, Iter: Iterator<Item = char>> {
    source: Iter,
    position: usize,
    context: &'a Context<'a>,
}

impl<'a> Input<'a, std::str::Chars<'a>> {
    fn new(source: &'a str, context: &'a Context<'a>) -> Self {
        Self {
            source: source.chars(),
            position: 0,
            context,
        }
    }

    fn eos(&self) -> (&'a Context<'a>, std::ops::Range<usize>) {
        (self.context, 0..0)
    }
}

impl<'a, Iter: Iterator<Item = char>> Iterator for Input<'a, Iter> {
    type Item = (char, (&'a Context<'a>, std::ops::Range<usize>));

    fn next(&mut self) -> Option<Self::Item> {
        match self.source.next() {
            Some(c) => {
                self.position += 1;
                Some((c, (self.context, self.position..(self.position + 1))))
            }
            None => None,
        }
    }
}

fn render_error_report<'a>(error: &ErrorType<'a>, with_color: bool) -> Report<SpanType<'a>> {
    let builder = Report::build(
        ReportKind::Error,
        error.span().context(),
        error.span().start(),
    );

    let builder = builder.with_config(Config::default().with_color(with_color));

    // TODO(OmarTawfik): also disable colors below if with_color is false

    let msg = if let chumsky::error::SimpleReason::Custom(msg) = error.reason() {
        msg.clone()
    } else {
        format!(
            "{}{}, expected {}",
            if error.found().is_some() {
                "Unexpected token"
            } else {
                "Unexpected end of input"
            },
            if let Some(label) = error.label() {
                format!(" while parsing {}", label)
            } else {
                String::new()
            },
            if error.expected().len() == 0 {
                "something else".to_string()
            } else {
                error
                    .expected()
                    .map(|expected| match expected {
                        Some(expected) => expected.to_string(),
                        None => "end of input".to_string(),
                    })
                    .collect::<Vec<_>>()
                    .join(" or ")
            },
        )
    };

    let builder = builder.with_code(3).with_message(msg).with_label(
        Label::new(error.span())
            .with_message(match error.reason() {
                chumsky::error::SimpleReason::Custom(msg) => msg.clone(),
                _ => format!(
                    "Unexpected {}",
                    error
                        .found()
                        .map(|c| format!("token {}", c.fg(Color::Red)))
                        .unwrap_or_else(|| "end of input".to_string())
                ),
            })
            .with_color(Color::Red),
    );

    let builder = match error.reason() {
        chumsky::error::SimpleReason::Unclosed { span, delimiter } => builder.with_label(
            Label::new(span.clone())
                .with_message(format!(
                    "Unclosed delimiter {}",
                    delimiter.fg(Color::Yellow)
                ))
                .with_color(Color::Yellow),
        ),
        chumsky::error::SimpleReason::Unexpected => builder,
        chumsky::error::SimpleReason::Custom(_) => builder,
    };

    return builder.finish();
}
