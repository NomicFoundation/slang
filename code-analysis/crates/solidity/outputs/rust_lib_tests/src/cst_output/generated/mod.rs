// This file is generated via `cargo build`. Please don't edit by hand.

mod contract_definition;
mod source_unit;

use std::rc::Rc;

use anyhow::{bail, Result};
use solidity_rust_lib::generated::{
    cst,
    parse::{ParserType, Parsers},
};

pub const BREAKING_VERSIONS: [&str; 3] = ["0.4.10", "0.6.0", "0.8.13"];

pub fn get_parser<'a>(
    parsers: Parsers<'a>,
    parser_name: &str,
) -> Result<ParserType<'a, Rc<cst::Node>>> {
    return Ok(match parser_name {
        "contract_definition" => parsers.contract_definition,
        "source_unit" => parsers.source_unit,
        _ => bail!("Unrecognized parser_name: {parser_name}"),
    });
}
