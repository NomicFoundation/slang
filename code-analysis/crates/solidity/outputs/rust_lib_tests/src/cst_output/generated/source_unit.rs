use crate::cst_output::run;
use anyhow::Result;

#[test]
fn empty_file() -> Result<()> {
    return run("source_unit", "empty_file");
}
