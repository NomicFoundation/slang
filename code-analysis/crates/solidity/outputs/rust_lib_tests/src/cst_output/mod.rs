mod generated;

use std::io::Write;
use std::rc::Rc;

use anyhow::Result;
use insta::assert_snapshot;
use semver::Version;
use solidity_rust_lib::generated::cst;
use solidity_rust_lib::generated::parse::Parsers;
use solidity_rust_lib::internal_api::parse;
use solidity_rust_lib::internal_api::ParserOutput;

use crate::cst_output::generated::{get_parser, BREAKING_VERSIONS};

pub fn run(parser_name: &str, test_name: &str) -> Result<()> {
    let manifest_dir = std::env::var("CARGO_MANIFEST_DIR")?;
    let test_dir = std::path::PathBuf::from(manifest_dir)
        .join("snapshots/cst_output")
        .join(parser_name)
        .join(test_name);

    let input_file = test_dir.join("input.sol");
    let input = std::fs::read_to_string(&input_file)?;

    let mut last_snapshot: String = String::new();

    for version in BREAKING_VERSIONS {
        let version = Version::parse(version)?;
        let parsers = Parsers::new(&version);
        let parser = get_parser(parsers, parser_name)?;
        let output = parse(&input, parser, /* with_color */ false);
        let current_snapshot = render_output(&output)?;

        if current_snapshot == last_snapshot {
            // Skip versions that don't change its output.
            continue;
        }

        let mut settings = insta::Settings::clone_current();
        settings.set_omit_expression(true);
        settings.set_input_file(&input_file);
        settings.set_snapshot_path(test_dir.join("output"));
        settings.set_snapshot_suffix(format!("{parser_name}@{test_name}@{version}"));
        settings.set_prepend_module_to_snapshot(false);

        settings.bind(|| assert_snapshot!(current_snapshot));

        last_snapshot = current_snapshot;
    }

    return Ok(());
}

fn render_output(output: &ParserOutput<Rc<cst::Node>>) -> Result<String> {
    let mut snapshot = Vec::new();

    if !output.error_reports.is_empty() {
        for report in &output.error_reports {
            write!(&mut snapshot, "{report}")?;
        }

        writeln!(&mut snapshot, "---")?;
    }

    if let Some(root_node) = &output.root_node {
        let ron_tree = serde_yaml::to_string(&root_node)?;
        write!(&mut snapshot, "{ron_tree}")?;
        writeln!(&mut snapshot, "---")?;
    }

    return Ok(String::from_utf8(snapshot)?);
}
