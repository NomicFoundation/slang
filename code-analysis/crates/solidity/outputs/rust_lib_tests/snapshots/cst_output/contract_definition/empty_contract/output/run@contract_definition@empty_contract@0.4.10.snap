---
source: crates/solidity/outputs/rust_lib_tests/src/cst_output/mod.rs
input_file: crates/solidity/outputs/rust_lib_tests/tests/cst_output/contract_definition/empty_contract/input.sol
---
!Rule
kind: ContractDefinition
children:
- None
- !Token
  kind: Contract
  range:
    start: 0
    end: 8
  trivia:
  - !Rule
    kind: LeadingTrivia
- !Token
  kind: RawIdentifier
  range:
    start: 9
    end: 15
  trivia:
  - !Rule
    kind: LeadingTrivia
    children:
    - !Token
      kind: Whitespace
      range:
        start: 8
        end: 9
- None
- !Rule
  kind: DelimitedContractBodyElements
  children:
  - !Token
    kind: OpenBrace
    range:
      start: 16
      end: 17
    trivia:
    - !Rule
      kind: LeadingTrivia
      children:
      - !Token
        kind: Whitespace
        range:
          start: 15
          end: 16
  - !Rule
    kind: ContractBodyElements
  - !Token
    kind: CloseBrace
    range:
      start: 17
      end: 18
    trivia:
    - !Rule
      kind: LeadingTrivia
    - !Rule
      kind: TrailingTrivia
      children:
      - None
      - !Token
        kind: EndOfLine
        range:
          start: 18
          end: 19
---

